launcher/electron/app/main.js

```diff
@@ -0,0 +1,156 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+var app = require('app');  // Module to control application life.
+var ipc = require('ipc');
+var http = require('http');
+var crashReporter = require('crash-reporter');
+
+var events = require('events');
+var backendRunner = require('./backend-runner.js');
+var mainMenu = require('./main-menu.js');
+var windowManager = require('./window-manager.js');
+
+var appReady = false;
+var openFile;
+var mainMenu;
+
+// Report crashes to our server.
+crashReporter.start();
+
+// Electron ready
+app.on('ready', function() {
+  // Run beaker backend
+  backendRunner.startNew().on('ready', connectToBackend);
+});
+
+// Kill backend before exiting
+app.on('quit', function() {
+  killBackend();
+});
+
+// Fired when OS opens file with application
+app.on('open-file', function(event, path) {
+  event.preventDefault();
+  if (appReady) {
+    windowManager.newWindow(backendRunner.getUrl() + '/beaker/#/open?uri=' + path)
+  } else {
+    openFile = path;
+  }
+});
+
+ipc.on('quit', function() {
+  app.quit();
+});
+
+mainMenu.on('quit', function() {
+  app.quit();
+});
+
+ipc.on('try-change-server', function() {
+  windowManager.newWindow('file://' + __dirname + '/templates/change-server-dialog.html', 'popup');
+});
+
+mainMenu.on('try-change-server', function() {
+  windowManager.newWindow('file://' + __dirname + '/templates/change-server-dialog.html', 'popup');
+});
+
+ipc.on('change-server', function(e, address) {
+  switchToBackend(address);
+});
+
+mainMenu.on('change-server', function(e, address) {
+  switchToBackend(address);
+});
+
+ipc.on('new-backend', function() {
+  killBackend();
+  backendRunner.startNew().on('ready', connectToBackend);
+});
+
+mainMenu.on('new-backend', function() {
+  killBackend();
+  backendRunner.startNew().on('ready', connectToBackend);
+});
+
+mainMenu.on('new-empty-notebook', function() {
+  windowManager.newWindow(backendRunner.getUrl() + 'beaker/#/session/empty');
+});
+
+mainMenu.on('new-default-notebook', function() {
+  windowManager.newWindow(backendRunner.getUrl() + 'beaker/#/session/new');
+});
+
+ipc.on('new-window', function(e, url) {
+  windowManager.newWindow(url);
+});
+
+function switchToBackend(address) {
+  if (address != backendRunner.getUrl()) {
+    killBackend();
+  }
+  // Open new control panel there
+  console.log('Switching to ' + address);
+  windowManager.newWindow(address);
+  backendRunner.setUrl(address);
+  backendRunner.setLocal(false);
+}
+
+function connectToBackend() {
+  // Have to wait until backend is fully ready
+  spinUntilReady(backendRunner.getHash() + '/beaker/rest/util/ready', function() {
+    windowManager.connectToBackend();
+    // Open file if launched with file
+    if (openFile !== undefined) {
```

Comment:
we should use 
`typeof openFile !== 'undefined'`

since ^ will fail in strict mode